router/peers_test.go

```diff
@@ -0,0 +1,170 @@
+}
+
+func checkBlank(t *testing.T, update []byte) {
+	decoder := gob.NewDecoder(bytes.NewReader(update))
+	peerInfo := decodePeerInfo(t, decoder)
+	if len(peerInfo) != 0 {
+		t.Fatalf("%s: Expected 0 items but got %s", wt.CallSite(2), peerInfo)
+	}
+}
+
+func checkEncoding(t *testing.T, update []byte, routers []*Router, connections [][]Connection) {
+	decoder := gob.NewDecoder(bytes.NewReader(update))
+
+	// Peers can come either way round, so read them all in and sort them
+	peerInfo := decodePeerInfo(t, decoder)
+	sort.Sort(byName(peerInfo))
+	N := len(peerInfo)
+	if N != len(routers) {
+		t.Fatalf("%s: Expected %d items but got %d: %s", wt.CallSite(2), len(routers), N, peerInfo)
+	}
+	for i := 0; i < N; i++ {
+		checkPeerDetails(t, peerInfo[i], routers[i].Ourself.Peer)
+		checkConnsEncoding(t, peerInfo[i].name, peerInfo[i].connsBuf, connections[i])
+	}
+}
+
+func TestPeersEncoding(t *testing.T) {
+	const (
+		peer1NameString = "01:00:00:01:00:00"
+		peer2NameString = "02:00:00:02:00:00"
+		peer3NameString = "03:00:00:03:00:00"
+	)
+	var (
+		peer1Name, _ = PeerNameFromString(peer1NameString)
+		peer2Name, _ = PeerNameFromString(peer2NameString)
+		peer3Name, _ = PeerNameFromString(peer3NameString)
+	)
+
+	// Create some peers
+	r1 := NewTestRouter(t, peer1Name, nil)
+	r2 := NewTestRouter(t, peer2Name, nil)
+	r3 := NewTestRouter(t, peer3Name, nil)
+
+	// Check state when they have no connections
+	checkEncoding(t, r1.Peers.EncodeAllPeers(), rs(r1), ca(nil))
+	checkEncoding(t, r2.Peers.EncodeAllPeers(), rs(r2), ca(nil))
+
+	// Now try adding some connections
+	r1.AddTestConnection(r2)
+	r2.AddTestConnection(r1)
+	checkEncoding(t, r1.Peers.EncodeAllPeers(), rs(r1, r2), ca(cs(r2), nil))
+	checkEncoding(t, r2.Peers.EncodeAllPeers(), rs(r1, r2), ca(nil, cs(r1)))
+	// Currently, the connection from 2 to 3 is one-way only
+	r2.AddTestConnection(r3)
+	checkEncoding(t, r2.Peers.EncodeAllPeers(), rs(r1, r2, r3), ca(nil, cs(r1, r3), nil))
+	checkEncoding(t, r3.Peers.EncodeAllPeers(), rs(r3), ca(nil))
+}
+
+func TestPeersGarbageCollection(t *testing.T) {
+	const (
+		peer1NameString = "01:00:00:01:00:00"
+		peer2NameString = "02:00:00:02:00:00"
+		peer3NameString = "03:00:00:03:00:00"
+	)
+	var (
+		peer1Name, _ = PeerNameFromString(peer1NameString)
+		peer2Name, _ = PeerNameFromString(peer2NameString)
+		peer3Name, _ = PeerNameFromString(peer3NameString)
+	)
+
+	// Create some peers with some connections to each other
+	r1 := NewTestRouter(t, peer1Name, nil)
+	r2 := NewTestRouter(t, peer2Name, nil)
+	r3 := NewTestRouter(t, peer3Name, nil)
+	r1.AddTestConnection(r2)
+	r2.AddTestRemoteConnection(r1, r2)
+	r2.AddTestConnection(r1)
+	r2.AddTestConnection(r3)
+	r3.AddTestConnection(r1)
+	r1.AddTestConnection(r3)
+	r2.AddTestRemoteConnection(r1, r3)
+	r2.AddTestRemoteConnection(r3, r1)
+
+	// Drop the connection from 2 to 3, and 3 isn't garbage-collected because 1 has a connection to 3
+	r2.DeleteTestConnection(r3)
+	peersRemoved := r2.Peers.GarbageCollect()
```

Comment:
instead of instrumenting the gc with a different callback, a better test would to be check whether Peers contain exactly the peers we expect.